### PR TITLE
feat(core,phrases): validate social sign-in method with social connector IDs

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

Validate the social sign-in method and social connector IDs

… → #480 → #485 → …

## Linear Issue Reference

LOG-2069

## Testing

image

### CODE REVIEW

File: packages/core/src/lib/sign-in-experience.test.ts [hunk 1]
```diff
@@ -122,4 +130,50 @@ describe('validate sign-in methods', () => {
...
+        );
+      }).toMatchError(new RequestError('sign_in_experiences.invalid_social_connectors'));
+    });
+
+    test('should not throw when all selected social connectors are enabled and social sign-in method is enabled', () => {
+      expect(() => {
+        validateSignInMethods(
+          { ...mockSignInMethods, social: SignInMethodState.secondary },
+          ['facebook', 'github'],
+          // @ts-expect-error-this-line
```
Comment (wangsijie) [inline packages/core/src/lib/sign-in-experience.test.ts:173]: try `as unknown as xxx` to match type(in tests)

**Review by wangsijie — APPROVED** [review]

File: packages/core/src/lib/sign-in-experience.test.ts (hunk 1 above)
Comment (IceHe) [reply packages/core/src/lib/sign-in-experience.test.ts]: Updated to `as ConnectorInstance[]`. PTAL.

**Comment by IceHe** [comment]
Rebased to resolve conflicts.

File: packages/core/src/lib/sign-in-experience.ts [hunk 2]
```diff
@@ -24,8 +25,33 @@ export const validateTermsOfUse = (termsOfUse: TermsOfUse) => {
...
+  const enabledSocialConnectorIds = new Set(
+    enabledConnectorInstances
+      .filter((instance) => instance.metadata.type === ConnectorType.Social)
+      .map((instance) => instance.connector.id)
+  );
+  assertThat(
+    socialSignInConnectorIds.every((id) => enabledSocialConnectorIds.has(id)),
+    'sign_in_experiences.invalid_social_connectors'
+  );
+};
```
Comment (simeng-li) [inline packages/core/src/lib/sign-in-experience.ts:50]: nit: could be a general one, `assertConnectorsEnabled` which regardless of the connector type, for whatever id provided, check the existence and the enable status.

File: packages/core/src/lib/sign-in-experience.ts [hunk 3]
```diff
@@ -56,7 +82,7 @@ export const validateSignInMethods = (
       'sign_in_experiences.enabled_connector_not_found',
       { type: ConnectorType.Social }
     );
-    // TODO: assertNonemptySocialConnectorIds
-    // TODO: assertEnabledSocialConnectorIds
+    assertNonemptySocialConnectorIds(socialSignInConnectorIds);
+    assertEnabledSocialConnectorIds(socialSignInConnectorIds, enabledConnectorInstances);
```
Comment (simeng-li) [inline packages/core/src/lib/sign-in-experience.ts:86]: nit: not necessarily extract to a separate method.

**Review by simeng-li — APPROVED** [review]
overall LGTM

File: packages/core/src/lib/sign-in-experience.ts (hunk 3 above)
Comment (IceHe) [reply packages/core/src/lib/sign-in-experience.ts:86]: I'm good with both.

- If inlined, this function will be a little longer: 44 lines.
- Even though it won't be reused in another place, I think it's more readable to extract it and give it a good name.

Comment (IceHe) [reply packages/core/src/lib/sign-in-experience.ts]: Inlined.

**Pull request merged**